### DISCUSSION TITLE
Best practise for reacting to a transfered file

### DISCUSSION THREAD

**Discussion by waltkind** [arendst/Tasmota#22380] [Support request] [answered]
When my doorbell rings, I capture a picture and send it via ftp to my Lanbon L8.
Now I'm looking for the best way to trigger further action.
Now it works that the doorbell also triggers a mqtt device that is subscribed. Then I wait some seconds until the picture is converted and transfered before it is displayed.
I can check the existance of the file by< "path.exists('Door.png')==true" but how can I trigger the action as soon as the file is there.

**Comment by sfromis** [comment]
If by "ftp" you mean the FTP server you can enable with the "UFSFTP" command, that does not provide any signals you can pick up, when a file has been stored. Hence you could start polling for the file (with reasonably short intervals) after getting the MQTT message. Or you could have the doorbell end publish another message after having transferred the file.

If you instead use the Berry example FTP server, you could easily modify that to do something when a file was received and stored. https://github.com/arendst/Tasmota/blob/development/tasmota/berry/drivers/ftp.be

**Comment by waltkind** [comment]
Thank you for the input. I did not know about the Berry FTP server and perhaps that is a promissing way.

What you called polling was the way I tried but did not succeed.

```berry
.
   def t() print("warte auf Bild") end
  while path.exists('Door.png')!=true tasmota.set_timer(100,t) end
.
``` 
This code did not work. There must be some fault.
Is there better code for polling?

**Comment by sfromis** [comment, marked as answer]
That code is blocking, meaning that you cannot do anything (including transferring files) while it runs. It will not wait 100 ms between checks, but it will schedule a shitload of those timers, which does not exactly help the situation.

Something like this works better, doing the check inside the periodic checking function
```
def poller(max, found)
  import path
  if path.exists('Door.png')
    found()
  elif max
    print(f'Waiting...{max}')
    tasmota.set_timer(100,/->poller(max-1, found))
  else
    raise 'error','failed to receive file'
  end
end
poller(120, /->print('got it'))
```
Also illustrates how to proceed with further code when the file is found, and a sanity check to prevent infinite polling.

**Reply by waltkind** [reply]
Thank you, that works.
